crates/pathfinder/src/storage/contract.rs

```diff
@@ -69,6 +70,23 @@ impl ContractCodeTable {
         Ok(())
     }
 
+    pub fn update_declared_on_if_null(
+        transaction: &Transaction<'_>,
+        class: ClassHash,
+        block: StarknetBlockHash,
+    ) -> anyhow::Result<bool> {
+        let rows_changed = transaction.execute(
+            "UPDATE contract_code SET declared_on=? WHERE hash=? AND declared_on IS NULL",
+            rusqlite::params![block, class],
+        )?;
```

Comment:
Unfortunately this null check needs to check the database each time, since both declare and deploy can potentially be "declarations".